server/src/app.ts

```diff
@@ -82,17 +78,33 @@ export const main = async (app: Express) => {
           },
         },
       },
+      update: {},
     });
   }
 
+  async function findOrCreateUser(email: string) {
+    try {
+      return await upsertUser(email);
+    } catch (e) {
+      if (
+        e instanceof Prisma.PrismaClientKnownRequestError &&
+        e.code === 'P2002'
+      ) {
+        console.log(`user with ${email} already exists`);
+        return await upsertUser(email);
+      }
+      throw e;
+    }
```

Comment:
I'm a bit confused about the overall logic here.

If there's error thrown during the first `upsertUser` call and the error code is `P2002`, `upsertUser` is called again. Won't that just throw error once again?

Reply:
That's a very good question. The answer is 'no', but it's also not necessary to call `upsert` again. Reason being, `upsert` throws if two requests come in at the same time and resolve to two `createUser` calls. That can happen because there's a race between `findUser` and `createUser` (effectively that's what `upsert` is doing). If both `findUser`s resolve before the first `createUser`, then the second `createUser` will throw.

Since, the `upsert` is throwing because the `createUser` part has already resolved, we can just log that something weird happened and move on.
```suggestion
    try {
      return await upsertUser(email);
    } catch (e) {
      if (
        e instanceof Prisma.PrismaClientKnownRequestError &&
        e.code === 'P2002'
      ) {
        console.log(`user with ${email} already exists`);
      } else {
        throw e;
      }
    }
```

Reply:
> That can happen because there's a race between findUser and createUser (effectively that's what upsert is doing). If both findUsers resolve before the first createUser, then the second createUser will throw.

:100: 

@gikf [Here's an explanation from the Prisma docs](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#cause) on why the error occurs.